GoMeet/src/main/java/com/google/sps/data/AutoProposeTimes.java

```diff
@@ -0,0 +1,180 @@
+    } 
+  }
+
+  /**
+   * Generates a List of TimePeriod representing free times with a duration of 
+   * at least the value of 'meetingDuration'.
+   * @return A List of TimePeriod that are free periods where a
+   * meeting may be scheduled.
+   */
+  public List<TimePeriod> proposeTimes() 
+      throws IOException, GoogleJsonResponseException, GeneralSecurityException {
+    // Gather freebusy information for each of the calendars.
+    List<TimePeriod> busyPeriods = new ArrayList<TimePeriod>();
+    List<TimePeriod> freePeriods = new ArrayList<TimePeriod>();
+    for (int i = 0; i < calendarId.size(); i++) {
+      busyPeriods.addAll(freebusyRequest(calendarId.get(i)));
+    }
+
+    // If there are no busyPeriods, add a free time that is the whole period, and return
+    if (busyPeriods.size() == 0) {
+      freePeriods.add(newTimePeriod(this.startTime, this.endTime));
+      return freePeriods;
+    }
+
+    // Sort all the busy periods by start time: O(Nlog(N))
+    Collections.sort(busyPeriods, new TimePeriodComparator());
+
+    long durationStartOfDay;
+    // Add the free time block from startTime to first busy period, if any.
+    DateTime firstBusyPeriod = busyPeriods.get(0).getStart();
+    durationStartOfDay = firstBusyPeriod.getValue() - this.startTime.getValue();
+    if (durationStartOfDay >= this.meetingDuration) {
+      freePeriods.add(newTimePeriod(this.startTime, firstBusyPeriod));
+    }
+
+    // Generate the free periods: O(N)
+    int i = 0;
+    TimePeriod current = busyPeriods.get(i);
+    DateTime lastBusyPeriodEnd = current.getEnd();
+
+    while (i < busyPeriods.size()) {
+      int j = i + 1;
+      while (j < busyPeriods.size() 
+          && current.getEnd().getValue() > 
+              busyPeriods.get(j).getStart().getValue()) {
+        // While the end of the current event overlaps over the start of the next event
+        j++;
+      }
+      if (j == busyPeriods.size()) {
+        break;
+      }
+      // Check that the duration of the free period is at least
+      // of 'meetingDuration' length.
+      DateTime periodStart = current.getEnd(); 
+      DateTime periodEnd = busyPeriods.get(j).getStart();
+      long duration = periodEnd.getValue() - periodStart.getValue();
+      if (duration >= this.meetingDuration) {
+        freePeriods.add(newTimePeriod(periodStart, periodEnd));
+      }
+      // Start from the last TimePeriod we passed in inner loop.
+      i = j;
+      current = busyPeriods.get(i);
+      lastBusyPeriodEnd = current.getEnd();
+    }
+
+    // Add the free time block from end of lastest busy period to endTime, if any.
+    long durationEndOfDay;
+    durationEndOfDay = this.endTime.getValue() - lastBusyPeriodEnd.getValue();
+    if (durationEndOfDay >= this.meetingDuration) {
+      freePeriods.add(newTimePeriod(lastBusyPeriodEnd, this.endTime));
+    }
+
+    return freePeriods;
+  }
+
+  /**
+   * Makes a request to the Freebusy library of the Google Calendar API,
```

Comment:
With making requests to an external API, is there a chance that the request could not be processed? And should that be handled?

This may not be so important because I understand this is more to demonstrate the algorithm for finding free periods so just something to think about.

Reply:
Yep, I've handled the case (see [PR 51](https://github.com/googleinterns/step271-2020/pull/51)) where there are security error responses from the API. There will be an error thrown (GeneralSecurityError or GoogleJsonResponseException) by these functions that call the API, and I'm planning to have the servlet that calls these functions handle that by returning an error response to the user like we usually do.

I'm sure there are other errors (invalid request format, etc.) that I haven't explicitly handled, so I think I might modify my servlet class to just catch all exceptions and return the error message in an error response to the front-end. Thanks for your suggestion